k-distribution/include/kframework/builtin/domains.md

```diff
@@ -612,6 +612,94 @@ that two different sets will return the same element, even if they are similar.
 endmodule
 ```
 
+### Implementation of Sets
+
+The following lemmas are simplifications that the Haskell backend can
+apply to simplify expressions of sort `Set`.
+
+```k
+module SET-KORE-SYMBOLIC [kore,symbolic]
+  imports SET
+  imports private K-EQUAL
+  imports private BOOL
+
+  //Temporarly rule for #Ceil simplification, should be generated in front-end
+
+// Matching for this version not implemented.
+  // rule #Ceil(@S1:Set @S2:Set) =>
+  //        {intersectSet(@S1, @S2) #Equals .Set} #And #Ceil(@S1) #And #Ceil(@S2)
+  //   [simplification]
+
+//simpler version
+  rule #Ceil(@S:Set SetItem(@E:KItem)) =>
+         {(@E in @S) #Equals false} #And #Ceil(@S) #And #Ceil(@E)
+    [simplification]
+
+  // -Set simplifications
+  rule S              -Set .Set           => S          [simplification]
+  rule .Set           -Set  _             => .Set       [simplification]
+  rule SetItem(X)     -Set (S SetItem(X)) => .Set
+                               ensures notBool (X in S) [simplification]
+  rule S              -Set (S SetItem(X)) => .Set
+                               ensures notBool (X in S) [simplification]
+  rule (S SetItem(X)) -Set S              => SetItem(X)
+                               ensures notBool (X in S) [simplification]
+  rule (S SetItem(X)) -Set SetItem(X)     => S
+                               ensures notBool (X in S) [simplification]
+  // rule SetItem(X)     -Set S              => SetItem(X)
+  //                            requires notBool (X in S)  [simplification]
+  // rule (S1 SetItem(X)) -Set (S2 SetItem(X))  => S1 -Set S2
+  //                             ensures notBool (X in S1)
+  //                             andBool notBool (X in S2) [simplification]
+
+
+
+  // |Set simplifications
+  rule S    |Set .Set => S    [simplification]
+  rule .Set |Set S    => S    [simplification]
+  rule S    |Set S    => S    [simplification]
+
+  rule (S SetItem(X)) |Set SetItem(X) => S SetItem(X)
+                             ensures notBool (X in S) [simplification]
+//  rule (S SetItem(X)) |Set S          => S SetItem(X)
+//                             ensures notBool (X in S) [simplification]
+  rule SetItem(X) |Set (S SetItem(X)) => S SetItem(X)
```

Comment:
Since this [PR](https://github.com/runtimeverification/k/pull/2831), you should be able to use the `comm` attribute on simplification rules.
But this requires that the syntax and the rule are annotated for safety reasons.
So in this case, you may need to modify _domains.md_ as well.

Reply:
Thanks, I will use `comm` where appropriate then.